### PR TITLE
Implement POM design pattern in `cardano-cli.hs`

### PR DESCRIPTION
**Pull request by Jimbo4350**
Following from #249 & #243

### CODE REVIEW

File: cardano-node/app/cardano-cli.hs [hunk 1]
```diff
@@ -76,21 +84,81 @@ main = do
     renderCliError = show
 
 data CLI = CLI
-  { protocol    :: Protocol
+  { totalConfig :: PartialCardanoConfiguration
```
Comment (deepfire) [inline cardano-node/app/cardano-cli.hs]: Hmm, `totalConfig` vs.  `Partial*` type? : -)

What is our picture here?  Maybe we want a better name?

File: cardano-node/app/cardano-cli.hs [hunk 2]
```diff
@@ -76,21 +84,81 @@ main = do
...
+  socketDir <- parseSocketDir
+  pbftSigThresh <- parsePbftSigThreshold
+  reqNetMagic <- parseRequireNetworkMagic
+  slotLength <- parseSlotLength
+  logConfigFp <- lastOption parseLogConfigFile
+  logMetrics <- parseLogMetrics
+  pure $ CLI (createPcc ptcl dbPath genPath genHash delCert sKey socketDir
+                        pbftSigThresh reqNetMagic slotLength logConfigFp
+                        logMetrics)
+              cmd
```
Comment (deepfire) [inline cardano-node/app/cardano-cli.hs]: Don't you feel that with the `ApplicativeDo`'s introduction of the unnecessary names we've lost some of the clarity here?

I think previously it was much easier on the eyes..

File: cardano-node/src/Cardano/Common/Parsers.hs [hunk 3]
```diff
@@ -13,8 +13,6 @@ module Cardano.Common.Parsers
   , parseProtocolMockPBFT
   , parseProtocolPraos
   , parseProtocolRealPBFT
-  , parseProtocolActual
-  , parseProtocolAsCommand
```
Comment (deepfire) [inline cardano-node/src/Cardano/Common/Parsers.hs]: :+1:

File: cardano-node/app/cardano-cli.hs (hunk 2 above)
Comment (Jimbo4350) [reply cardano-node/app/cardano-cli.hs]: @deepfire I'm not opposed to removing `ApplicativeDo`. You have a good point.

File: cardano-node/app/cardano-cli.hs (hunk 1 above)
Comment (Jimbo4350) [reply cardano-node/app/cardano-cli.hs]: @deepfire Yep, this was a poor name choice. I'll change it.

File: cardano-node/app/cardano-cli.hs (hunk 2 above)
Comment (Jimbo4350) [reply cardano-node/app/cardano-cli.hs]: @deepfire Thinking about it some more, the issue here is then I would need `CLI Protocol FilePath...` etc. `ApplicativeDo` allows me to keep the `CLI` datatype more concise.

Comment (deepfire) [reply cardano-node/app/cardano-cli.hs]: Just pushed an example of how to do it without any type changes: https://github.com/input-output-hk/cardano-node/commit/a86716311c0d74860140fe61beda91a4565e4da1#diff-deb1de1e23655df9067b46df271864c6L109

**Pull request merged**